### PR TITLE
Application gets stuck when deleting last block of the same round twice - Closes #2032

### LINKED ISSUE #2032: Application gets stuck when deleting last block of the same round twice
### Expected behavior
Node should be able to delete the last block of the same round as many time as needed without corrupting snapshots
### Actual behavior
We perform snapshotting process at the second last block of the round here: https://github.com/LiskHQ/lisk/blob/development/modules/rounds.js#L228Is there any particular reason why we do this at the second last block? and not before at the last block of the round (before making any calculation for the next round?)
### Steps to reproduce:
```
describe('snapshotting when deleting last block of the round twice', () => {
	before(() => {
		// Forge 201 blocks to reach height 202 (genesis block is already tyhere)
		return Promise.mapSeries([...Array(201)], () => {
			return addTransactionsAndForgePromise(library, [], 0);
		}).then(() => {
			return deleteLastBlockPromise(library);
		}).then(() => {
			return addTransactionsAndForgePromise(library, [], 0);
		});
	});

	it('should be able to delete the block again', () => {
		return deleteLastBlockPromise(library);
	});
});
```
### Which version(s) does this affect? (Environment, OS, etc...)
1.0.0

### PR DESCRIPTION
**Pull request by 4miners**
### What was the problem?
Unable to delete last block of round twice in a row because round snapshot was not available.
### How did I fix it?
- Keep `mem_round_snapshot` and `mem_votes_snapshot` tables after round rollback, so those tables can be reused for same round.
- Check if there is snapshot available for current round before we restore it, to prevent usage of wrong snapshot (results with corrupted data) when going back more than one round of blocks.

It's not the best solution, but not so hard to implement and reliable. 
### How to test it?
Run test suite.
### Review checklist

* The PR solves #2032
* All new code is covered with unit tests
* All new code was formatted with Prettier
* Linting passes
* Tests pass
* Commit messages follow the [commit guidelines](CONTRIBUTING.md#git-commit-messages)
* Documentation has been added/updated

### CODE REVIEW

**Comment by 4miners** [comment]
Notice: This PR contains changes from https://github.com/LiskHQ/lisk/pull/2038 and https://github.com/LiskHQ/lisk/pull/2039, those should be merged prior to this one.

File: db/sql/rounds/check_snapshot_availability.sql [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  PARAMETERS: round - Round for we are checking availability
+*/
+
+SELECT (
+	CASE WHEN (
+		SELECT 1 AS available FROM mem_round_snapshot WHERE round = ${round} LIMIT 1
+	) = 1 THEN TRUE
+	ELSE (
+		CASE WHEN (
+			SELECT COUNT(*) FROM mem_round_snapshot
```
Comment (SargeKhan) [inline db/sql/rounds/check_snapshot_availability.sql]: This case of query is for when we rollback to first block?

Comment (SargeKhan) [reply db/sql/rounds/check_snapshot_availability.sql]: wouldn't it be nicer to just write to mem_round_snapshot and mem_votes_snapshot after first block?

File: test/functional/system/rounds.js [hunk 2]
```diff
@@ -726,6 +726,33 @@ describe('rounds', () => {
...
+				return addTransactionsAndForgePromise(library, [], 0);
+			});
+
+			it('should be able to delete last block of round again', () => {
+				return deleteLastBlockPromise();
+			});
+
+			// FIXME: Unskip that test after https://github.com/LiskHQ/lisk/issues/1781 is closed
+			// eslint-disable-next-line
+			it.skip('mem_accounts table should be equal to one generated before last block of round deletion', () => {
```
Comment (SargeKhan) [inline test/functional/system/rounds.js]: Issue #1781 is already closed.

File: test/functional/system/rounds.js [hunk 3]
```diff
@@ -1155,4 +1182,22 @@ describe('rounds', () => {
 			return expect(lastBlock.height).to.equal(101);
 		});
 	});
+
+	describe('deleting last block of round twice in a row - no transactions during round', () => {
+		before(() => {
+			return Promise.mapSeries([...Array(202)], () => {
```
Comment (SargeKhan) [inline test/functional/system/rounds.js]: This should be 201 since we already have genesisBlock in the database.

File: test/functional/system/rounds.js (hunk 2 above)
Comment (4miners) [reply test/functional/system/rounds.js]: On `beta.8` is not skipped, but in `beta.7` we have exactly same expectation that is still skipped.

File: db/sql/rounds/check_snapshot_availability.sql (hunk 1 above)
Comment (SargeKhan) [reply db/sql/rounds/check_snapshot_availability.sql]: I think this query is overcomplicated. We are adding this case specifically for the first round. So, we can modify the function that calls this query to handle that edge case (for first round), IMO we shouldn't move this logic to the database layer.

File: test/functional/system/rounds.js (hunk 3 above)
Comment (SargeKhan) [reply test/functional/system/rounds.js]: :/. I don't like these tests. If in the future it fails, the person looking at the code will have to understand all the tests to see what is the state of the chain (how many blocks it contains) before he can see what the actual problem is. Besides, it makes our tests tightly coupled with each other, meaning even just changing the order will make tests fail.

File: db/sql/rounds/check_snapshot_availability.sql (hunk 1 above)
Comment (4miners) [reply db/sql/rounds/check_snapshot_availability.sql]: It's not for the first round, it can happen in every round. There are 2 options:
- one query like this, that returns `true` or `false` if we can use snapshot
- 2 queries
-- one that give us info if snapshot table contains records for requested round
-- second that counts records in snapshot table
and perform condition in round logic:
```
	checkSnapshotAvailability() {
		return this.t.rounds
			.isRoundExistsInRoundSnapshot(this.scope.round)
			.then(isExists => {
				if (!isExists) {
					return this.t.rounds
						.countRoundSnapshot()
						.then(count => {
							if (count) {
								throw new Error(
									`Snapshot for round ${
										this.scope.round
										} not available, unable to perform round rollback`
								);
							}
						});
				}
			});
	}
```
For me first option is better because we have:
- less queries in db repo (don't need additional file + additional function + unit tests for it)
- less complicated logic and unit tests for checkSnapshotAvailability function

Comment (shuse2) [reply db/sql/rounds/check_snapshot_availability.sql]: As you said,
> It's not the best solution, but not so hard to implement and reliable.
this might not be final solution, so this might not be applicable,
- complicated SQL is harder to optimize the performance later
- being able / easy to test in application is better than harder to test on DB
- 1 complex query vs 2 simple query, I think 2 simple query would be better

File: test/functional/system/rounds.js (hunk 3 above)
Comment (4miners) [reply test/functional/system/rounds.js]: Yes, such a behavior is intended. Those tests are functional system, the longer chain is - the better reliability of those tests is.

File: db/sql/rounds/check_snapshot_availability.sql (hunk 1 above)
Comment (4miners) [reply db/sql/rounds/check_snapshot_availability.sql]: - There are only 2 simple conditions in that query, is it really complicated?
- It's not hard to test that query on DB, tests for that are part of this PR
- Regarding performance - it should be faster than executing 2 queries, and I don't think that there is a room for further performance optimization later.

Comment (4miners) [reply db/sql/rounds/check_snapshot_availability.sql]: By `not the best solution` I meant there are some better that require a little brainstorming, for example:
- decouple `tick` to `beforeTick` and `afterTick` (so we can snapshot before last block of round, not after one block before last block of round)
- rewrite round 😉

File: test/functional/system/rounds.js (hunk 3 above)
Comment (SargeKhan) [reply test/functional/system/rounds.js]: The purpose of these tests is not to check how the application behaves for the longer chain, they are for checking what happens when we delete the last block of the round twice. But I am not asking to change these tests, just talking about the structure we follow sometime for writing tests, so we may have a consistent style of writing tests by different developers. :)

File: db/sql/rounds/check_snapshot_availability.sql (hunk 1 above)
Comment (shuse2) [reply db/sql/rounds/check_snapshot_availability.sql]: I guess I'm just strongly against putting the logic in the database layer.
- 2 simple select can be optimize a lot easier than 1 sql with condition.
- it's always a lot harder to optimize database layer compare to application layer

Even though this 1 SQL would be performance wise slightly better, for consistency through out the code and future scalability, I think we should put it on application layer.

Comment (4miners) [reply db/sql/rounds/check_snapshot_availability.sql]: @shuse2 Ok, I will rewrite.

File: test/functional/system/rounds.js (hunk 3 above)
Comment (4miners) [reply test/functional/system/rounds.js]: @SargeKhan I can refactor those tests and decouple them to few separate files (file per scenario tested), if we all agree that they will be easier to maintain then. However after `beta.7` and `beta.8` get merged together, as those branches contain separate changes to those tests.

File: db/sql/rounds/check_snapshot_availability.sql (hunk 1 above)
Comment (nazarhussain) [reply db/sql/rounds/check_snapshot_availability.sql]: You guys are thinking this query complicated? You had not seen others queries already. 

Complicated SQL is a relative term, completely depending on the nature of the application you are dealing with. If you are writing a simple e-commerce system, then yes any query more than 2/2 lines or with sub queries can be considered complicated. But if you are dealing with some analytical system, normally a query with 2-3 joins and sub queries considered simple ;) 

Depending on the nature of the application we are dealing, we have to gave some luxury to database layer to handle some complexity. Don't consider DB layer as abstraction to run one line queries.  Testing DB level complex queries is not difficult at all, we already have tests for more complicated db scenarios for this purpose.

Comment (shuse2) [reply db/sql/rounds/check_snapshot_availability.sql]: `SELECT COUNT(*) FROM mem_round_snapshot WHERE round = ?`
can't you just do this and handle the error properly?

File: test/functional/system/rounds.js (hunk 3 above)
Comment (MaciejBaj) [reply test/functional/system/rounds.js]: For simplicity of merging let's leave them in the same file for now.

File: db/sql/rounds/check_snapshot_availability.sql (hunk 1 above)
Comment (MaciejBaj) [reply db/sql/rounds/check_snapshot_availability.sql]: I would suggest split the query to keep the logic in JS code where possible.

Comment (4miners) [reply db/sql/rounds/check_snapshot_availability.sql]: No, it will not work, we need to allow restoring snapshot in 2 cases:
- when we have some records with requested round (means we have data for that round, and can use it)
- when we don't have any records for requested round and table is empty (table is empty only when there was no transactions during entire round)

**Comment by 4miners** [comment]
@nazarhussain Done.
